main/pipeline/pipeline.h

```diff
@@ -22,17 +22,17 @@ std::vector<ast::ParsedFile> index(std::unique_ptr<core::GlobalState> &gs, std::
                                    const options::Options &opts, WorkerPool &workers,
                                    std::unique_ptr<KeyValueStore> &kvstore);
 
-std::vector<ast::ParsedFile> resolve(std::unique_ptr<core::GlobalState> &gs, std::vector<ast::ParsedFile> what,
-                                     const options::Options &opts, WorkerPool &workers, bool skipConfigatron = false);
+ast::MaybeASTPassResult resolve(std::unique_ptr<core::GlobalState> &gs, std::vector<ast::ParsedFile> what,
+                                const options::Options &opts, WorkerPool &workers, bool skipConfigatron = false);
 
 std::vector<ast::ParsedFile> incrementalResolve(core::GlobalState &gs, std::vector<ast::ParsedFile> what,
                                                 const options::Options &opts);
 
 std::vector<ast::ParsedFile> name(core::GlobalState &gs, std::vector<ast::ParsedFile> what,
                                   const options::Options &opts, bool skipConfigatron = false);
 
-std::vector<ast::ParsedFile> typecheck(std::unique_ptr<core::GlobalState> &gs, std::vector<ast::ParsedFile> what,
-                                       const options::Options &opts, WorkerPool &workers);
+ast::MaybeASTPassResult typecheck(std::unique_ptr<core::GlobalState> &gs, std::vector<ast::ParsedFile> what,
```

Comment:
It was kind of confusing to see a data structure called `ASTPass` be returned from `typecheck` which i don't think of as operating on the AST (it converts the AST to a CFG and then operates on that).

I might suggest naming your data structure something that hints that this is not specific to the core pipeline but in fact an artifact of LSP preemption.

Reply:
Do you have a suggestion? The data type is `ast::something-or-other`, so I thought it's an AST.

Would `MaybeAST` work? `MaybeASTOrCFG`?

Reply:
Mentioned elsewhere but I think a name that mentions LSP would be more appropriate, because whether this is null or non-null is 100% controlled by the bit in global state that's set by LSP.